File: 106.ConstructBinaryTreeFromInorderAndPostOrder.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {

    TreeNode build(Map<Integer, Integer> inIndexes, int[] postorder, int iLb, int iRb, int pLb, int pRb) {
        if (iLb > iRb) {
            return null;
        }

        TreeNode root = new TreeNode(postorder[pRb]);
        int mid = inIndexes.get(root.val);

        root.left = build(inIndexes, postorder, iLb, mid - 1, pLb, pLb + mid - iLb - 1);
        root.right = build(inIndexes, postorder, mid + 1, iRb, pRb - (iRb - mid), pRb - 1);
        return root;

    }

    public TreeNode buildTree(int[] inorder, int[] postorder) {
        Map<Integer, Integer> inIndexes = new HashMap<>();
        for(int i = 0; i < inorder.length; i++) {
            inIndexes.put(inorder[i], i);
        }
        return build(inIndexes, postorder, 0, inorder.length - 1, 0, postorder.length - 1);
    }
}

File: 106.ConstructBinaryTreeFromInorderAndPostOrder.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:

    def build(self, inRef, postorder, ilb, irb, plb, prb):
        if ilb > irb:
            return None

        root = TreeNode(postorder[prb])
        mid = inRef[root.val]

        root.left = self.build(inRef, postorder, ilb, mid - 1, plb, plb + mid - 1 - ilb)
        root.right = self.build(inRef, postorder, mid + 1, irb, prb - (irb - mid), prb - 1)
        
        return root

    def buildTree(self, inorder: List[int], postorder: List[int]) -> Optional[TreeNode]:
        inRef = {}
        for i in range(len(inorder)):
            inRef[inorder[i]] = i
        return self.build(inRef, postorder, 0, len(inorder) - 1, 0, len(postorder) - 1)
